### ISSUE TITLE
Error trying to migrate bot

### ISSUE THREAD

**Issue by bpleao** [Microsoft/BotBuilder#4227]
## Bot Info
* SDK Platform: Node.js
* SDK Version: 3.14.0
* Active Channels: Webchat, Direct Line, Skype
* Deployment Environment: Azure App Service

## Issue Description
I'm trying to migrate the bot to the new Azure service, but after clicking the migrate button and starting the process I get the following error message and the process stops:
An error occurred: 500
An internal error occurred at the server.

**Comment by EricDahlvang** [comment]
Do you have access to multiple azure subscriptions?  We have a workaround on our troubleshooting guide that might help: https://docs.microsoft.com/en-us/bot-framework/bot-service-troubleshoot-general-problems#why-cant-i-migrate-my-bot  Please review, and let us know if this workaround resolves the migration issue.

**Comment by bpleao** [comment]
@EricDahlvang, thanks, but I've tried the solution and it did not work. Only the error code changed from 500 to 400. Not sure if I've followed the guide exactly as described though. I don't know what "From the target directory" means in the instructions. I've added a guest user through Azure Active Directory, provided him with the proper roles and tried to migrate the bot using this user. Any further advice?

**Comment by EricDahlvang** [comment]
The target directory is the directory to which the bot is being migrated, not the directory the bot is currently in.

**Comment by bpleao** [comment]
Ok, but then how do I know what is the target directory and, more important, how do I add a new user from this target directory as stated in the instructions?

**Comment by EricDahlvang** [comment]
Go to https://aad.portal.azure.com and in the top right, choose the directory you are migrating the bot into (that is the target directory).  Then on the left, choose Users and add a new user:

![image](https://user-images.githubusercontent.com/11055362/36802120-e1bf6120-1c68-11e8-80d6-bf82fbdd54c8.png)

**Comment by bpleao** [comment]
I have no option to change directory in the top right:
![image](https://user-images.githubusercontent.com/12271581/36813147-6a6cf71e-1cb2-11e8-8b55-af6921cb8be3.png)

**Comment by EricDahlvang** [comment]
This means that your user has access to just the one azure tenant.  Do you see error details in the browser console?  Also, please share your MicrosoftAppId and BotHandle, and I will investigate further.

**Comment by bpleao** [comment]
Indeed there are error details in browser console. It seems to be 17 repetitions of the same error. Here is the log:
[dev.botframework.com-1519903368390.log](https://github.com/Microsoft/BotBuilder/files/1770935/dev.botframework.com-1519903368390.log)
Microsoft App Id: e513659a-768f-4c0b-89cb-dd5b46d42d60
Bot Handle: consolador

Thanks!

**Comment by bpleao** [comment]
@EricDahlvang, any news on this topic? Thanks.

**Comment by Andrea-Orimoto** [comment]
Hi @bpleao, sorry about the inconvenience, we're seeing the issue in our logs and we're investigating. Will let you know as soon as we understand what's causing this.

**Comment by bpleao** [comment]
Hi @Andrea-Orimoto, any news on this?

**Comment by bpleao** [comment]
Hi @EricDahlvang, any news on this?

**Comment by Andrea-Orimoto** [comment]
Hi @bpleao, sorry about the delay, no news yet. If we can't unblock you in time, we will whitelist your bot until we fix your issue. Will be in touch soon.

**Comment by Andrea-Orimoto** [comment]
Hi @bpleao, can you contact me at andreo@microsoft.com, so we can discuss how to unblock you? thanks and sorry for the inconvenience.

**Comment by bpleao** [comment]
Hi @Andrea-Orimoto, just sent you an email.  Thanks

**Comment by Fujihita** [comment]
Hi, I'm also having the same issue migrating my bot to the new Azure service. When I clicked migrate button, I got the following error (code 400):

> The operation is not permitted for namespace 'Microsoft.BotService'. List of permitted provider namespaces is 'Microsoft.Authorization,Microsoft.Features,microsoft.insights,Microsoft.NotificationHubs,Microsoft.Resources,Microsoft.Sql,microsoft.support,microsoft.visualstudio,Microsoft.Web,SuccessBricks.ClearDB'.

I also have only access to one azure tenant and the bot I have has the same specs as bpleao's bot. Should I send Andrea an email too?

**Comment by EricDahlvang** [comment]
@Fujihita If you are still experiencing a problem with migration, please open a new issue. https://github.com/Microsoft/BotBuilder/issues/new

Closing this, as Andrea-Orimioto is handling it via email.  Thanks Andrea!

**Comment by Andrea-Orimoto** [comment]
This is still open, we've put this on hold, but still looking at it. I'll follow up on this one.

**Comment by maba4891** [comment]
I have the same problem as @bpleao, sent an email to @Andrea-Orimoto with the details.

**Comment by Andrea-Orimoto** [comment]
@Fujihita still having problems? if yes, please create a new issue and provide information about your bot.

**Comment by Andrea-Orimoto** [comment]
@bpleao, I finally found the issue(s). The solution first:

1.  First of all, use [FuncPack](https://github.com/Azure/azure-functions-pack) to pack your code and node_modules. This is highly recommended for Node.js bots running on functions, as noted in our [migration doc](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-migrate-bot).
1.a before packing, change module.exports = { default: withLogging } to module.exports = withLogging; in your index.js
1.b all the .json files you're loading, should wither be copied to the new .funcpack folder after packing, or you should use "require" for them, so that funcpack packs them as well (I believe)

2. Once your current bot is packed and working, then migrate it, using a new destination name. The dialog will suggest you consolador-upgrade, change it to something else. If you re-use a previously used name, it will fail.

Now a brief description of the issues:
1. App Service name re-use. If you run migration multiple times, if you re-use a name, then migration will fail to get the functions secrets. --> solution: do not re-use the same name on subsequent migrations.
2. Failure to npm install post migration. Once a bot is migrated, migration should kick off npm install. There are two issues here. a) It fails to kick off or b) it takes a while to complete. In both cases the bot does not work, prompting you to rollback. --> solution: use FuncPack so you don't need npm install

Hope this is clear, let me know if you have any questions.
Andrea

**Comment by Andrea-Orimoto** [comment]
One more important step I just discovered. Once you move to funcpack, delete the node_modules folder from the app service prior to migration. A function using funcpack does not need the node_modules anymore, and that folder might cause the timeout during migration.

**Comment by Andrea-Orimoto** [comment]
hi @bpleao, were you able to try this? Let me know if you have any questions.

**Comment by bpleao** [comment]
@Andrea-Orimoto: sorry, I haven't been able to try it yet. I'll do that asap and I'll let you know how it goes. Thanks.
